plugins/GSdx/Renderers/Vulkan/GSDeviceVK.cpp

```diff
@@ -0,0 +1,238 @@
+	int i = 0;
+	bool queueIndicesFound = false;
+	for (const auto& queueFamily : queueFamilies) {
+		VkBool32 presentSupport = false;
+		vkGetPhysicalDeviceSurfaceSupportKHR(m_vk.physicalDevice, i, m_vk.surface, &presentSupport); // Assumes our current device supports presentation. Could cause issues.
+		if (queueFamily.queueCount > 0 && queueFamily.queueFlags & VK_QUEUE_GRAPHICS_BIT) {
+			m_vk.graphicsFamily = i;
+		}
+		if (queueFamily.queueCount > 0 && presentSupport) {
+			m_vk.presentFamily = i;
+		}
+		if (m_vk.graphicsFamily != std::numeric_limits<uint32_t>::max() && m_vk.presentFamily != std::numeric_limits<uint32_t>::max()) {
+			queueIndicesFound = true;
+			break;
+		}
+		i++;
+	}
+	if (!queueIndicesFound) {
+		fprintf(stderr, "Vulkan: Couldn't get the required queue family indices\n");
+		return false;
+	}
+	std::vector<VkDeviceQueueCreateInfo> queueCreateInfos;
+	std::set<uint32_t> uniqueQueueFamilies = {
+		m_vk.graphicsFamily, m_vk.presentFamily
+	};
+	float queuePrirority = 1.0f;
+	for (uint32_t queueFamily : uniqueQueueFamilies) {
+		VkDeviceQueueCreateInfo queueCreateInfo = {};
+		queueCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+		queueCreateInfo.queueFamilyIndex = m_vk.graphicsFamily;
+		queueCreateInfo.queueCount = 1;
+		queueCreateInfo.pQueuePriorities = &queuePrirority;
+		queueCreateInfos.push_back(queueCreateInfo);
+	}
+	VkPhysicalDeviceFeatures deviceFeatures = {};
+	VkDeviceCreateInfo deviceCreateInfo = {};
+	deviceCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
+	deviceCreateInfo.pQueueCreateInfos = queueCreateInfos.data();
+	deviceCreateInfo.queueCreateInfoCount = static_cast<uint32_t>(queueCreateInfos.size());
+	deviceCreateInfo.pEnabledFeatures = &deviceFeatures;
+	deviceCreateInfo.enabledExtensionCount = static_cast<uint32_t>(m_vk.deviceExtensions.size());
+	deviceCreateInfo.ppEnabledExtensionNames = m_vk.deviceExtensions.data();
+	deviceCreateInfo.enabledLayerCount = 0;
+	result = vkCreateDevice(m_vk.physicalDevice, &deviceCreateInfo, nullptr, &m_vk.logicalDevice);
+	if (result != VK_SUCCESS) {
+		fprintf(stderr, "Vulkan: Couldn't create the logical device\n");
+		return false;
+	}
+
+	return true;
+}
+
+VkPhysicalDevice GSDeviceVK::PickDevice(std::vector<VkPhysicalDevice>& devices)
+{
+	VkPhysicalDevice out = VK_NULL_HANDLE;
+	for (VkPhysicalDevice dev : devices) {
+		uint32_t eCount;
+		vkEnumerateDeviceExtensionProperties(dev, nullptr, &eCount, nullptr);
+		std::vector<VkExtensionProperties> availableExtensions(eCount);
+		vkEnumerateDeviceExtensionProperties(dev, nullptr, &eCount, availableExtensions.data());
+		std::set<std::string> requiredExtensions(m_vk.deviceExtensions.begin(), m_vk.deviceExtensions.end());
+		for (const auto& extension: availableExtensions) {
+			requiredExtensions.erase(extension.extensionName);
+		}
+		if (requiredExtensions.empty()) {
+				// TODO choose best gpu or let users choose which GPU to use.
```

Comment:
honestly let's do device stuff in the next PR. that way we can do it properly at once. it will require some annoying changes.